aws/middleware/middleware.go

```diff
@@ -166,3 +180,45 @@ func AddRawResponseToMetadata(stack *middleware.Stack) error {
 func GetRawResponse(metadata middleware.Metadata) interface{} {
 	return metadata.Get(rawResponseKey{})
 }
+
+// AddTimeOffsetBuildMiddleware sets a value representing clock skew on the request context.
+// This can be read by other operations (such as signing) to correct the date value they send
+// on the request
+type AddTimeOffsetBuildMiddleware struct {
+	Offset *atomic.Int64
+}
+
+// ID the identifier for AddTimeOffsetBuildMiddleware
+func (m *AddTimeOffsetBuildMiddleware) ID() string { return "AddTimeOffsetMiddleware" }
+
+// HandleBuild sets a value for attemptSkew on the request context if one is set on the client.
+func (m AddTimeOffsetBuildMiddleware) HandleBuild(ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler) (
+	out middleware.BuildOutput, metadata middleware.Metadata, err error,
+) {
+	if m.Offset != nil {
+		offset := time.Duration(m.Offset.Load())
+		ctx = SetAttemptSkewContext(ctx, offset)
+	}
+	return next.HandleBuild(ctx, in)
+}
+
+// AddTimeOffsetDeserializeMiddleware sets the clock skew on the client if it's present on the context
+// at the end of the request
+type AddTimeOffsetDeserializeMiddleware struct {
+	Offset *atomic.Int64
+}
+
+// ID the identifier for AddTimeOffsetDeserializeMiddleware
+func (m *AddTimeOffsetDeserializeMiddleware) ID() string { return "AddTimeOffsetDeserializeMiddleware" }
+
+// HandleDeserialize gets the clock skew context from the context, and if set, sets it on the pointer
+// held by AddTimeOffsetDeserializeMiddleware
+func (m *AddTimeOffsetDeserializeMiddleware) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+	out middleware.DeserializeOutput, metadata middleware.Metadata, err error,
+) {
+	v := GetAttemptSkewContext(ctx)
+	if v != 0 {
+		m.Offset.Store(v.Nanoseconds())
```

Comment:
Discussed this offline. We could only store this once when we see it, on the assumption that you only need to collect skew once instead of on every call. This could free some of the calls to `lock` that `atomic` is likely doing behind the scenes. 

However, the downside is that if we ever get a bad `Date` back from the server, it can end up poisoning the time for all subsequent requests. 

Decided it's probably safer to update it every time